### PR TITLE
Falcon40b enable multi-device tensors

### PR DESCRIPTION
**Pull request by djordje-tt**
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/9171

### Problem description
- Ops are running in for loop for every device, instead using multi-device tensors for loop is moved to c++, e2e perf is significantly improved and code is much cleaner.

### What's changed
- All for loops are replaced with single call of op and operands are single tensors which are multi-device and parts of it are placed on different devices.
- Changes made: 
  - Tensor creation now uses mullti-device approach
  - For loops removed and operands are tensors spread/cloned across devices

### Checklist
- [x] Post commit CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/9584021621
- [x] Demo T3K tests: https://github.com/tenstorrent/tt-metal/actions/runs/9584036485
- [x] Unit T3K tests: https://github.com/tenstorrent/tt-metal/actions/runs/9594062241/job/26456036271 (f40b tests passed ttnn tests fail for timeout - same as on main)
- [x] Frequent T3K tests: https://github.com/tenstorrent/tt-metal/actions/runs/9594066507/job/26456028433 - llama fails
- [x] Model perf T3K tests: https://github.com/tenstorrent/tt-metal/actions/runs/9594085277

### CODE REVIEW

**Comment by djordje-tt** [comment]
Please @johanna-rock-tt, @uaydonat, @s-jovic, while reviewing changes ignore all optimized files, they will be removed as soon as all pipelines pass.

File: models/demos/t3000/falcon40b/tt/falcon_mlp.py [hunk 1]
```diff
@@ -108,84 +77,73 @@ def __call__(
...
+            output_mem_config=self.model_config["DENSE_H_TO_4H_MM_OUTPUT_MEMCFG"],
+            output_dtype=self.model_config["DENSE_H_TO_4H_MM_OUTPUT_DTYPE"],
+            compute_kernel_config=self.model_config["COMPUTE_KERNEL_CONFIG"],
+        )
+        x.deallocate(True)
+
+        hidden_states = ttnn.experimental.tensor.sharded_to_interleaved(
+            hidden_states, output_mem_config=self.model_config["DEFAULT_MEMCFG"]
+        )
         hidden_states = ttnn.all_gather(
```
Comment (djordje-tt) [inline models/demos/t3000/falcon40b/tt/falcon_mlp.py:94]: @ayerofieiev-tt, here is one of examples of all_gather for multi-device tensors.

**Review by s-jovic — APPROVED** [review]
Gtm as long as the CI runs pass

File: models/demos/t3000/falcon40b/tests/test_falcon_attention.py [hunk 2]
```diff
@@ -359,7 +342,8 @@ def test_FalconAttention_inference(
     model_config_str,
     model_location_generator,
     get_tt_cache_path,
-    all_devices,
+    # all_devices,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_attention.py]: remove comment

File: models/demos/t3000/falcon40b/tests/test_falcon_attention.py [hunk 3]
```diff
@@ -262,10 +243,12 @@ def run_test_FalconAttention_inference(
         use_cache=use_cache,
     )
 
-    tt_out = torch.cat([tt2torch_tensor(tt_o).squeeze(1) for tt_o in tt_out], -1)
+    tt_out = ttnn.to_torch(tt_out, device=device_mesh, mesh_composer=ConcatMeshToTensor(device_mesh, dim=3))
+
+    # torch.cat([tt2torch_tensor(tt_o).squeeze(1) for tt_o in tt_out], -1)
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_attention.py]: remove comment

File: models/demos/t3000/falcon40b/tests/test_falcon_causallm.py [hunk 4]
```diff
@@ -351,7 +341,7 @@ def test_FalconCausalLM_inference(
 
     input_shape = [batch, seq_len]
     model_config = get_model_config(model_config_str, llm_mode, input_shape, num_devices)
-    devices = get_devices_for_t3000(all_devices, num_devices)
+    devices = t3k_device_mesh.get_devices()
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_causallm.py:344]: We should remove `num_devices` as argument and use the number of devices in the `t3k_device_mesh` to pass into `get_model_config` instead.
Same for every other test.

File: models/demos/t3000/falcon40b/tests/test_falcon_decoder.py [hunk 5]
```diff
@@ -86,74 +87,58 @@ def run_test_FalconDecoder_inference(
...
-                )
-            )
+        tt_attention_mask = ttnn.as_tensor(
+            tensor=attention_mask_bool,
+            dtype=model_config["ATTN_MASK_DTYPE"],
+            layout=ttnn.TILE_LAYOUT,
+            device=device_mesh,
+            memory_config=attention_mask_memconfig,
+            mesh_mapper=ShardTensorToMesh(device_mesh, dim=1),
+            preprocess=lambda x: (x * -1e5).expand(-1, device_mesh.get_num_devices(), -1, -1),  # TODO: TRY 8 here
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tests/test_falcon_decoder.py:112]: Forgot to remove

Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_decoder.py]: try 8 here?

File: models/demos/t3000/falcon40b/tests/test_falcon_mlp.py [hunk 6]
```diff
@@ -139,7 +143,8 @@ def test_FalconMLP_inference(
     model_config_str,
     model_location_generator,
     get_tt_cache_path,
-    all_devices,
+    # all_devices,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_mlp.py:146]: remove

File: models/demos/t3000/falcon40b/tests/test_falcon_mlp.py [hunk 7]
```diff
@@ -149,17 +154,17 @@ def test_FalconMLP_inference(
...
     if compute_grid_size.x < model_config["MAX_GRID_SIZE"][0] or compute_grid_size.y < model_config["MAX_GRID_SIZE"][1]:
         pytest.skip(f"Requires grid size of at least {model_config['MAX_GRID_SIZE']} to run")
 
     tt_cache_path = get_tt_cache_path(
         model_version, model_subdir="Falcon", default_dir=model_config["DEFAULT_CACHE_PATH"]
     )
-
     run_test_FalconMLP_inference(
-        devices,
+        # devices,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_falcon_mlp.py:166]: remove

File: models/demos/t3000/falcon40b/tests/test_perf_e2e_falcon.py [hunk 8]
```diff
@@ -238,8 +237,8 @@ def run_test_FalconCausalLM_end_to_end(
                 use_cache=use_cache,
             )
             tt_outs.append(tt_out)
-        tt_outs = [[tt_o.cpu() for tt_o in tt_out] for tt_out in tt_outs]
-
+        # Return single tile to synchronize device (sync_device on all 8 devices is slow)
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_perf_e2e_falcon.py:240]: did you create an issue for this yet?

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 9]
```diff
@@ -179,7 +171,8 @@ def run_test_FalconCausalLM_end_to_end(
             layer_past_len=kv_cache_len,
             use_cache=use_cache,
         )
-        tt_out = [tt_o.cpu() for tt_o in tt_out]
+        tt_out = ttnn.to_torch(tt_out, device=device_mesh, mesh_composer=ConcatMeshToTensor(device_mesh, dim=-1))
+        # tt_out = [tt_o.cpu() for tt_o in tt_out]
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:175]: remove

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 10]
```diff
@@ -232,7 +230,11 @@ def run_test_FalconCausalLM_end_to_end(
                 layer_past_len=kv_cache_len,
                 use_cache=use_cache,
             )
-            tt_out = [tt_o.cpu() for tt_o in tt_out]
+            tt_outs = ttnn.to_torch(tt_out, device=device_mesh, mesh_composer=ConcatMeshToTensor(device_mesh, dim=-1))
+            # tt_out = [tt_o.cpu() for tt_o in tt_out]
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:234]: remove

File: models/demos/t3000/falcon40b/tt/falcon_attention.py [hunk 11]
```diff
@@ -387,210 +333,174 @@ def fwd_prefill(
...
+        attn_weights.deallocate(True)
 
         return attn_output_slice
 
     def fwd_decode(
         self,
         hidden_states: ttnn.experimental.tensor.Tensor,
         alibi: torch.Tensor,
-        attention_mask: torch.Tensor,
+        attention_mask: ttnn.experimental.tensor.Tensor,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_attention.py:503]: can we use ttnn.Tensor instead?

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py [hunk 12]
```diff
@@ -32,17 +33,17 @@ def __init__(
         self.hidden_size = config.hidden_size
         self.state_dict = state_dict
         self.base_url = base_url
-        self.devices = devices
+        self.device_mesh = device_mesh
         self.layer_num = layer_num
         self.max_position_embeddings = max_position_embeddings
         self.model_config = model_config
-        self.num_devices = len(devices)
+        self.num_devices = len(device_mesh.get_device_ids())
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_decoder.py:40]: I think we don't need num_devices anymore, right? please check here and also in all other modules where we had it before, remove if we don't use it

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py [hunk 13]
```diff
@@ -298,36 +265,29 @@ def fwd_prefill(
...
-                memory_config=self.model_config["DROPOUT_ADD_OUTPUT_MEMCFG"],
-                output_tensor=output[i],
-            )
-
-            mlp_output[i].deallocate(True)
+        output = ttnn.add(
+            output,
+            mlp_output,
+            memory_config=self.model_config["DROPOUT_ADD_OUTPUT_MEMCFG"],
+            # output_tensor=output,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_decoder.py:288]: remove

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py [hunk 14]
```diff
@@ -298,36 +265,29 @@ def fwd_prefill(
...
-                    memory_config=self.model_config["PARALLEL_ATTN_ADD_OUTPUT_MEMCFG"],
-                    output_tensor=residual[i],
-                )
-            )
-            attention_output[i].deallocate(True)
+        output = ttnn.add(
+            output,
+            attention_output,
+            memory_config=self.model_config["PARALLEL_ATTN_ADD_OUTPUT_MEMCFG"],
+            # output_tensor=output,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_decoder.py:274]: remove

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py [hunk 15]
```diff
@@ -415,36 +364,29 @@ def fwd_decode(
...
-                memory_config=self.model_config["DROPOUT_ADD_OUTPUT_MEMCFG"],
-                output_tensor=output[i],
-            )
-
-            mlp_output[i].deallocate(True)
+        output = ttnn.add(
+            output,
+            mlp_output,
+            memory_config=self.model_config["DROPOUT_ADD_OUTPUT_MEMCFG"],
+            # output_tensor=output,
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_decoder.py:387]: remove and please check everywhere for commented code

File: models/demos/t3000/falcon40b/tt/falcon_model.py [hunk 16]
```diff
@@ -38,28 +40,28 @@ def __init__(
...
+        self.device_mesh = device_mesh
         self.state_dict = state_dict
         self.base_url = base_url
         self.config = config
         self.max_position_embeddings = max_position_embeddings
         self.model_config = model_config
         self.num_layers = num_layers
         self.hidden_size = config.hidden_size
-        self.num_devices = len(devices)
+        self.num_devices = device_mesh.get_num_devices()
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_model.py:51]: needed?

File: models/demos/t3000/falcon40b/tt/model_utils.py [hunk 17]
```diff
@@ -6,7 +6,8 @@
 import math
 
 import ttnn
-from models.utility_functions import torch2tt_tensor
+from ttnn import ReplicateTensorToMesh
+import ttnn.experimental
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/model_utils.py:10]: import ttnn?

**Review by johanna-rock-tt — APPROVED** [review]
A couple of things in my comments, otherwise looks good to me given all CI pipelines pass.

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py (hunk 13 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/falcon_decoder.py:288]: That is kept for a reason. ttnn.add is not working in_place with output_tensor=output, ttnn.add would be in_place so I kept it to uncomment once in_place ttnn.add is fixed for multi-device tensors.

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py (hunk 14 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/falcon_decoder.py:274]: Same as above

File: models/demos/t3000/falcon40b/tt/falcon_decoder.py (hunk 15 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/falcon_decoder.py:387]: Same as above

File: models/demos/t3000/falcon40b/tt/falcon_model.py (hunk 16 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/falcon_model.py:51]: Used down as dim for tensor creation.

File: models/demos/t3000/falcon40b/tt/model_utils.py (hunk 17 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/model_utils.py:10]: I removed all ttnn imports but "import ttnn"

File: models/demos/t3000/falcon40b/demo/demo.py [hunk 18]
```diff
@@ -415,7 +406,13 @@ def run_falcon_demo_kv(
             del tt_prefill_inputs
             del tt_prefill_attention_mask
 
-            logits = torch.cat([logits_torch.squeeze(1) for logits_torch in tt_tensors_to_torch_tensors(tt_logits)], -1)
+            # untilize data first
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/demo/demo.py:409]: It's not only for making sure device execution is finished. If you look for logits, they are later appended to prefill input and printed to console. Same applies for decode stage below. So if this is removed we wouldn't have proper output in demo script.

File: models/demos/t3000/falcon40b/demo/demo.py [hunk 19]
```diff
@@ -472,7 +469,15 @@ def run_falcon_demo_kv(
         if tt_decode_attention_mask is not None:
             del tt_decode_attention_mask
 
-        logits = torch.cat([logits_torch.squeeze(1) for logits_torch in tt_tensors_to_torch_tensors(tt_logits)], -1)
+        if tt_logits.get_layout() == ttnn.TILE_LAYOUT:
```
[the remark this answers is not in the document]
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/demo/demo.py:472]: Same as above.

File: models/demos/t3000/falcon40b/tests/test_falcon_causallm.py (hunk 4 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tests/test_falcon_causallm.py:344]: I agree but I would put a separate PR for that.

**Review by johanna-rock-tt — APPROVED** [review]
